### PR TITLE
gsoc: Add Action: Generate Patches

### PR DESCRIPTION
**Pull request by Nosferatul**
Add action that generates patches depending on the file

Closes https://github.com/coala/coala/issues/4599

### CODE REVIEW

**Comment by Nosferatul** [comment]
http://imgur.com/0VZ5Bhi

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import os
+from pygments.lexers import guess_lexer_for_filename
+from coalib.results.result_actions.ResultAction import ResultAction
```
Comment (satwikkansal) [inline coalib/results/result_actions/GeneratePatchesAction.py]: Space missing between imports :)

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+from coalib.results.result_actions.ShowPatchAction import ShowPatchAction
+from coalib.collecting.Collectors import (
+    collect_bears, collect_bears_by_aspects)
+from coalib.output.printers.LogPrinter import LogPrinter
+from coalib.parsing.FilterHelper import FilterHelper
+from coalib.parsing.DefaultArgParser import default_arg_parser
+from pyprint.ConsolePrinter import ConsolePrinter
```
Comment (satwikkansal) [inline coalib/results/result_actions/GeneratePatchesAction.py]: `pyprint` is a third party module, should not be mixed with local imports

File: coalib/settings/SectionFilling.py [hunk 3]
```diff
@@ -81,7 +81,6 @@ def fill_settings(sections,
     """
     local_bears = {}
     global_bears = {}
-
```
Comment (satwikkansal) [inline coalib/settings/SectionFilling.py]: Irrelevant change :)

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+        for i, action in enumerate(possible_options, 1):
+            show_possibilities(console_printer, i, action)
+
+        choose_action = int(input('[{:>5}]  Enter a number: '.format('')))
+        list_of_bears = list(possible_options.keys())
+
+        args.default_actions = '*: ApplyPatchAction'
+        bears_class.append(list_of_bears[choose_action - 1])
+        args.no_config = True
```
Comment (satwikkansal) [inline coalib/results/result_actions/GeneratePatchesAction.py]: It is already set to `True` in L51 https://github.com/coala/coala/pull/4634/files#diff-adce0bc484df7b1f0cf40333a8dfae54R51

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        possible_options = {}
+        args = default_arg_parser().parse_args()
+        args.files = []
+        log_printer = LogPrinter()
+        to_filename = None
+
+        for filename, this_diff in sorted(result.diffs.items()):
+            to_filename = this_diff.rename if this_diff.rename else filename
+            to_filename = '/dev/null' if this_diff.delete else to_filename
+            original_file_dict[filename]
```
Comment (satwikkansal) [inline coalib/results/result_actions/GeneratePatchesAction.py]: Why is this line required?

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        language = guess_lexer_for_filename(to_filename, 'Hello').name
+        path = to_filename
+        path = path.rstrip(os.sep)
+        os.path.basename(path)
+
+        args.no_config = True
+        filtered_bears = FilterHelper.get_filtered_bears(
+            'language', [language], None)
+        filtered_bears = filtered_bears[0]
```
Comment (satwikkansal) [inline coalib/results/result_actions/GeneratePatchesAction.py]: Isn't it better to do it this way

```py
filtered_bears = FilterHelper.get_filtered_bears(
           'language', [language], None)[0]['cli']
```

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        choose_action = str(input('[{:>4}]  Enter a number: '.format('')))
+        if choose_action is '' or choose_action is '0':
+            return False
+
+        choose_action = int(choose_action)
+        list_of_bears = possible_options
+
+        try:
+            chosed_bear = list_of_bears[choose_action - 1]
+        except IndexError:  # pragma: no cover
```
Comment (Nosferatul) [inline coalib/results/result_actions/GeneratePatchesAction.py]: This was added because of tests. In IRL it's impossible that `list_of_bears` is empty because with whatever file we run coala, always should find bears that can run on it because the user could choose an action.

Comment (yukiisbored) [inline coalib/results/result_actions/GeneratePatchesAction.py]: I think it would be nice to add an explanation for it.

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        filtered_bears = FilterHelper.apply_filters(
+            [['language', language]], None)[0]['cli']
+
+        filtered_bears2 = FilterHelper.apply_filters([['language', 'Python']])
+        print(filtered_bears2)
+
+        b = GBear('SpaceConsistencyBear')
+
+        filtered_bears = list(filtered_bears)
+        filtered_bears.insert(0, b)
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: `b` can inlined over here as it is not used anywhere else.

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class GBear():
+    def __init__(self, bear: str):
+        self.bear = bear
+        self.name = bear
+
+
+class GeneratePatchesAction(ResultAction):
+    SUCCESS_MESSAGE = 'Patch applied successfully.'
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: May be `Patch generated successfully.` ?

File: tests/output/ConsoleInteractionTest.py [hunk 10]
```diff
@@ -549,6 +553,43 @@ def test_ask_for_actions_and_apply5(self):
...
+                [action.get_metadata(), action2.get_metadata()],
+                {'ApplyPatchAction': action, 'GeneratePatchesAction': action2},
+                failed_actions, Result('origin', 'message'), {}, {}, {}]
+
+        with simulate_console_inputs('g', 5) as generator:
+            ask_for_action_and_apply(*args)
+            self.assertEqual(generator.last_input, 0)
+            self.assertNotIn('GeneratePatchAction', failed_actions)
+
+    def test_ask_for_action_generate2(self):
```
Comment (shreyans800755) [inline tests/output/ConsoleInteractionTest.py]: Can you please give some meaningful name to test case instead of just suffixing 2 to avoid confusion ? (or may be suffix something meaningful instead of just number)

File: tests/output/ConsoleInteractionTest.py [hunk 11]
```diff
@@ -549,6 +553,43 @@ def test_ask_for_actions_and_apply5(self):
             self.assertEqual(generator.last_input, 1)
             self.assertNotIn('ChainPatchAction', failed_actions)
 
+    def test_ask_for_action_generate(self):
+        failed_actions = set()
+        action = ApplyPatchAction()
+        action2 = GeneratePatchesAction()
```
Comment (shreyans800755) [inline tests/output/ConsoleInteractionTest.py]: Please use meaningful name for `action` and `action2` in both test cases.

File: coalib/results/result_actions/GeneratePatchesAction.py (hunk 8 above)
Comment (Nosferatul) [inline coalib/results/result_actions/GeneratePatchesAction.py]: hmm, it's used upper.

Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: @Nosferatul  Its just defined above, not used.
You can do `filtered_bears.insert(0, GBear('SpaceConsistencyBear'))`

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+            return False
+
+        choose_action = int(choose_action)
+        list_of_bears = possible_options
+
+        try:
+            chosed_bear = list_of_bears[choose_action - 1]
+        # In IRL it's impossible that `list_of_bears` is empty because
+        # that means the file used cannot be changed by any bear
+        except IndexError:  # pragma: no cover
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: If its impossible to get this exception, do we really need to catch it over here ?

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        console_printer.print('[{:>4}] *0. Do Nothing'.format(''))
+        for i, action in enumerate(possible_options, 1):
+            show_possibilities(console_printer, i, action)
+
+        choose_action = str(input('[{:>4}]  Enter a number: '.format('')))
+        if choose_action is '' or choose_action is '0':
+            return False
+
+        choose_action = int(choose_action)
+        list_of_bears = possible_options
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: I hope you know that this is just giving an alias to `possible_options` named `list_of_bears` and not actually copying whole list to different memory location.
I don't see a point of doing that as its just used once on line #83, that too only for read purpose and not updating the list.

File: coalib/results/result_actions/GeneratePatchesAction.py (hunk 12 above)
Comment (Nosferatul) [inline coalib/results/result_actions/GeneratePatchesAction.py]: yea, at some point I changed the code, apparently, now is redundant

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        section.add_or_create_setting(Setting('no_config', True))
+
+        console_printer = ConsolePrinter()
+        possible_options = []
+        log_printer = LogPrinter()
+        bears_app = []
+        args = default_arg_parser().parse_args()
+        args.files = []
+        to_filename = None
+        to_filename = sorted(result.diffs.items())[0][0]
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: are you redefining the same var from line 52?

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        language = guess_lexer_for_filename(to_filename, 'Hello').name
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: why hello?

Comment (Nosferatul) [inline coalib/results/result_actions/GeneratePatchesAction.py]: If it doesn't find any language for `to_filename` it should return something, I changed that to `Error, no file found` but it shouldn't happen because coala already found it when the user choosed this action.

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+def show_possibilities(console_printer, i, action):
+    console_printer.print('[{:>4}]  {}. Apply patch (\'{}\')'.format('', i,
+                                                                     action))
+
+
+class GBear():
+    def __init__(self, bear: str):
+        self.bear = bear
+        self.name = bear
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: wait. why do you have 2 members if both are equal to the same thing?

Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: are you modifying any? cuz i cant really see

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        choose_action = int(choose_action)
+        chosed_bear = possible_options[choose_action - 1]
+
+        args.default_actions = '*: ApplyPatchAction'
+
+        bears_app.append(chosed_bear)
+        args.files.append(os.path.basename(path))
+        args.bears = bears_app
+
+        mode_normal(console_printer, log_printer, args, debug=False)
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: this function is way too long. try to split it in helper functions (private methods). for example, initializing the lists could be an ieda (lines 46+), and so on.

Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: right now its not readable, it has no comments, probably only you understand the flow

Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: make documented smaller functions and its going to be better at reviewing

Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: also i know this isnt the case in other similar classes or so but please use constants. explain what a "-2" is at indexing (path_offset) or what a [[language, ..][0] is.. (that 0 can be a X_INDEX) or so

Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: because this way its really unmaintainable, if you want your code to be maintainable you just create a constant at the start and use it everywhere

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def show_possibilities(console_printer, i, action):
+    console_printer.print('[{:>4}]  {}. Apply patch (\'{}\')'.format('', i,
+                                                                     action))
+
+
+def define_arg(files, bears):
+    """
+    A function that generate a `default_arg_parser`
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: generates

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+OBJECT_INDEX = FILENAME_INDEX = 0
+DEFAULTBEAR = 'SpaceConsistencyBear'
+
+
+def show_possibilities(console_printer, i, action):
+    console_printer.print('[{:>4}]  {}. Apply patch (\'{}\')'.format('', i,
+                                                                     action))
+
+
+def define_arg(files, bears):
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: create_arg_parser feels like a better name

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+        choose_action = str(input('[{:>4}]  Enter a number: '.format('')))
+        if choose_action is '' or choose_action is '0':
+            return False
+
+        choose_action = int(choose_action)
+        chosed_bear = [possible_options[choose_action - 1]]
+
+        mode_normal(console_printer, log_printer, define_arg([
+            to_filename], chosed_bear), debug=False)
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/GeneratePatchesAction.py]: chosed doesnt exist in english, its chosen :)

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    A function that generates a `default_arg_parser`
+
+    :param files: A list that contains filenames.
+    :param bears: A list that contains name of bears.
+    :return:      An object of type `default_arg_parser`.
+    """
+    args = default_arg_parser().parse_args()
+    args.files = files
+    args.bears = bears
+    args.default_actions = '*: ApplyPatchAction'
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: `ApplyPatchAction` -> `GeneratePatchAction`

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+from coalib.results.result_actions.ResultAction import ResultAction
+from coalib.results.result_actions.ShowPatchAction import ShowPatchAction
+from coalib.output.printers.LogPrinter import LogPrinter
+from coalib.parsing.FilterHelper import FilterHelper
+from coalib.coala_modes import mode_normal
+from coalib.parsing.DefaultArgParser import default_arg_parser
+
+
+OBJECT_INDEX = FILENAME_INDEX = 0
+DEFAULTBEAR = 'SpaceConsistencyBear'
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: `DEFAULTBEAR` -> `DEFAULT_BEAR` for consistency.

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+
+class DefaultBear():
+    def __init__(self):
+        self.name = DEFAULTBEAR
+
+
+class GeneratePatchesAction(ResultAction):
+    SUCCESS_MESSAGE = 'Patch generated successfully.'
+
+    is_applicable = staticmethod(ShowPatchAction.is_applicable)
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: `ShowPatchAction` -> `GeneratePatchAction`
Copy past worm

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+                                                         'found').name
+
+        # Find all the bears that can run with the language found
+        filtered_bears = FilterHelper.apply_filters(
+            [['language', language]], None)[0]['cli']
+
+        filtered_bears = list(filtered_bears)
+        filtered_bears.insert(0, DefaultBear())
+        for b in filtered_bears:
+            possible_options.append(b.name)
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: `possible_options = [b.name for b in filtered_bears]`
You can also remove first definition of `possible_options = []` as its not being used in between.

File: coalib/results/result_actions/GeneratePatchesAction.py (hunk 21 above)
Comment (Nosferatul) [reply coalib/results/result_actions/GeneratePatchesAction.py]: you want to apply the patch, not to generate another one :)

File: coalib/results/result_actions/GeneratePatchesAction.py (hunk 23 above)
Comment (Nosferatul) [reply coalib/results/result_actions/GeneratePatchesAction.py]: This should not be changed, look in `ApplyPatchAction`

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+              file_diff_dict):
+        """
+        (G)enerate patches
+        """
+
+        console_printer = ConsolePrinter()
+        log_printer = LogPrinter()
+        to_filename = sorted(result.diffs.items())[OBJECT_INDEX][FILENAME_INDEX]
+
+        # Find the language used in the file provided by the user
```
Comment (yukiisbored) [inline coalib/results/result_actions/GeneratePatchesAction.py]: I personally think this code needs to be break into a number of functions.

Comment (Nosferatul) [reply coalib/results/result_actions/GeneratePatchesAction.py]: Hmm, sorry I don't understand what line exactly? 60-61 or 57?

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+        if choose_action is '' or choose_action is '0':
+            return False
+
+        choose_action = int(choose_action)
+        chosen_bear = [possible_options[choose_action - 1]]
+
+        mode_normal(console_printer, log_printer, create_arg_parser([
+            to_filename], chosen_bear), debug=False)
+
+        return True
```
Comment (shreyans800755) [inline coalib/results/result_actions/GeneratePatchesAction.py]: Shouldn't we return the value returned by `mode_normal` in the previous line ?

**Comment by Adrianzatreanu** [comment]
ack 1ee7feb

**Comment by Nosferatul** [comment]
@yukiisbored  do you have something to add ?

**Comment by Adrianzatreanu** [comment]
ack 25e019c

**Comment by Nosferatul** [comment]
@rultor merge
